### ISSUE TITLE
All nodes are busy or have too many pending compactions for the remaining candidate segments.

### ISSUE THREAD

**Issue by afsanchez83** [thelastpickle/cassandra-reaper#1069]
[Project board link](https://github.com/orgs/k8ssandra/projects/8/views/1?pane=issue&itemId=13937772)

**Spotted on 2.2.4 (da70e93)**

Hi,

I´m facing an issue after update to 2.2.4 version, this is my current configuration:

Datacenter_availability: EACH
Database_backend: postgres.

I have two datacenter for my Cassandra cluster and one reaper node for each datacenter hence I´m using Datacenter Availability "EACH", as I undertand the repaer should send the segments to repair only for their nodes in their datacenter but checking the logs I see this:

`INFO   [SchedulingManagerTimer] i.c.s.SegmentGenerator - Coalescing segments for nodes [10.164.208.100, 10.164.208.101, 10.164.208.102, 10.42.105.16, 10.42.105.27, 10.42.105.50]
INFO   [SchedulingManagerTimer] i.c.s.SegmentGenerator - Got enough tokens for one segment (67623345051890923) : [(7837481819336454,12806876652901556], (-4816065785917697407,-4813637698706516416], (6346952120928368691,6355886548882903170], (-7540423632381438062,-7530981842562336431], (617191263223489076,618102555379165687], (-5498906832766134728,-5496790966077876793], (-4829217124928932415,-4826902253705682213], (-4831832516149085216,-4829217124928932415], (-566783942501907762,-564495695134054679]]
INFO   [SchedulingManagerTimer] i.c.s.RepairManager - Starting a run with id #00000000-0000-0065-0000-000000000000 with current state 'NOT_STARTED'
INFO   [beatrixcluster:00000000-0000-0065-0000-000000000000] i.c.s.RepairRunner - Attempting to run new segment...
INFO   [beatrixcluster:00000000-0000-0065-0000-000000000000] i.c.s.RepairRunner - All nodes are busy or have too many pending compactions for the remaining candidate segments.
INFO   [beatrixcluster:00000000-0000-0065-0000-000000000000] i.c.s.RepairRunner - All nodes are busy or have too many pending compactions for the remaining candidate segments.
INFO   [beatrixcluster:00000000-0000-0065-0000-000000000000] i.c.s.RepairRunner - Repair amount done 0.0
INFO   [beatrixcluster:00000000-0000-0065-0000-000000000000] i.c.s.RepairRunner - Attempting to run new segment...
INFO   [beatrixcluster:00000000-0000-0065-0000-000000000000] i.c.s.RepairRunner - All nodes are busy or have too many pending compactions for the re`

I dont know why the reaper in 208 network is sending segments to the nodes in 105 network, and checking the repairs in progress I see that the only repair in progress for that cluster is waiting for available nodes.

![image](https://user-images.githubusercontent.com/79571496/118641794-978e3480-b7a8-11eb-8388-7c914a519d42.png)

Checking the compactionstats in the nodes inside the network 208 there are not pending tasks.

Do you know why it´s happening?.

Thanks in advance.

┆Issue is synchronized with this [Jira Story](https://datastax.jira.com/browse/REAP-109) by [Unito](https://www.unito.io)
┆Issue Number: REAP-109

**Comment by afsanchez83** [comment]
Repairs freeze and show this error in logs:

`INFO   [cluster:00000000-0000-00d4-0000-000000000000] i.c.s.RepairRunner - Attempting to run new segment...
INFO   [cluster:00000000-0000-00dc-0000-000000000000] i.c.s.RepairRunner - Attempting to run new segment...
INFO   [cluster:00000000-0000-00d4-0000-000000000000] i.c.s.RepairRunner - Next segment to run : 00000000-0001-451e-0000-000000000000
INFO   [cluster:00000000-0000-00dc-0000-000000000000] i.c.s.PostgresStorage - Failed to take lead on node 10.164.208.100 for segment 00000000-0001-451e-0000-000000000000
INFO   [cluster:00000000-0000-00dc-0000-000000000000] i.c.s.RepairRunner - Next segment to run : 00000000-0001-5266-0000-000000000000
INFO   [cluster:00000000-0000-00dc-0000-000000000000:00000000-0001-5266-0000-000000000000] i.c.j.JmxProxy - Triggering repair of range (8276362383583115617,8280376300688884691] for keyspace "keyspace" on host 10.164.208.100, with repair parallelism parallel, in cluster with Cassandra version '3.11.6' (can use DATACENTER_AWARE 'true'), for column fa`

Is there some issue to use repair parallelism "PARALLEL" in a datacenter availability "EACH"? it's a bit confising, thre repairs works from time to time, but some of them never start.

![image](https://user-images.githubusercontent.com/79571496/120329572-fd56e200-c2b9-11eb-858f-fd644fecb07b.png)

Last events for the 3 cases:
- Hanged: All nodes are busy or have too many pending compactions for the remaining candidate segments.
- Not started: Maximum parallel repairs reached, postponing repair run.
- Running: Triggered repair of segment 00000000-0001-51c3-0000-000000000000 via host 10.164.208.101

Do the repairs have any order or something for which some do not start and others do?

Thanks in advance.

**Comment by adejanovski** [comment]
Since 2.2.0, Reaper will allow concurrent segments from different repair runs to be running on the same nodes. To avoid overwhelming nodes, a new setting was introduced named `maxParallelRepairs` which defaults to `2`. This value is per repaired cluster.
In the current case, you already have 2 concurrent running repairs, so the others were put on hold. There's no(t yet) any prioritization between runs, and if a slot is available then any of the waiting repairs will start. I guess it could be good to select the oldest, but the current implementation doesn't allow this yet.

Your hanged repair is actually running, unless you have no RUNNING segment on it, which would mean there's another problem that's blocking it.

You could raise the `maxParallelRepairs` value to allow more concurrency if your cluster can handle it, to unblock your other repairs.

**Comment by adejanovski** [comment]
Now that I think of it, there could be some race conditions in `EACH` mode since there's no consensus on the parallel repairs limiter. One Reaper instance could have started repair B while the other started repair C, and at the same time be running A.
Could you switch your datacenterAvailability to `LOCAL` to see if you get better progress?

**Comment by adejanovski** [comment]
@afsanchez83, we've removed the postgres backend since 3.0 which will no longer get fixes or improvements. Would it be possible for you to switch to the Cassandra backend and assess if that problem still shows up?
